File: DataStructures/UnionFind/implementation/UnionFind.py
# Implementation of Union Find / Disjoint Set Union (DSU) in python

# DSU object will be instantiated and called as such:
# obj = UnionFind()
# obj.unite(x, y)
# parent = obj.findRootOf(x)
# connected = obj.isConnected(x, y)

class UnionFind:
    def __init__(self, n):
        ''' Create n sets '''
        self.root = [i for i in range(n)]
        self.rank = [1] * n
        self.sets = n

    def findRootOf(self, x):
        ''' Find the root of x '''
        while self.root[x] != x:
            self.root[x] = self.root[self.root[x]]
            x = self.root[x]
        return x

    def unite(self, x, y):
        ''' Unite the sets of x and y '''
        rootX = self.findRootOf(x)
        rootY = self.findRootOf(y)

        # Return if they belong from the same sets
        if rootX == rootY:
            return

        if self.rank[rootY] > self.rank[rootX]:
            self.root[rootX] = rootY
            self.rank[rootY] += self.rank[rootX]
        else:
            self.root[rootY] = rootX
            self.rank[rootX] += self.rank[rootY]

        self.sets -= 1

    def isConnected(self, x, y):
        ''' Check if x and y belong to the same set '''
        return self.findRootOf(x) == self.findRootOf(y)

File: DataStructures/UnionFind/implementation/UnionFind.cpp
// Implementation of Union Find / Disjoint Set Union (DSU) in C++

/*
 * DSU object will be instantiated and called as such:
 * UnionFind uf(n);
 * uf.unite(x, y);
 * int parent = uf.findRootOf(x);
 * bool connected = uf.isConnected(x, y);
*/

#include <vector>
using namespace std;

class UnionFind {
private:
    vector<int> root;
    vector<int> rank;
    int sets;

public:
    // Constructor to create n sets
    UnionFind(int n) {
        root.resize(n);
        rank.resize(n, 1);
        sets = n;
        for (int i = 0; i < n; ++i) {
            root[i] = i;
        }
    }

    // Find the root of x (with path compression)
    int findRootOf(int x) {
        if (root[x] != x) {
            root[x] = findRootOf(root[x]); // Path compression
        }
        return root[x];
    }

    // Unite the sets containing x and y
    void unite(int x, int y) {
        int rootX = findRootOf(x);
        int rootY = findRootOf(y);

        // If they are already in the same set, return
        if (rootX == rootY) {
            return;
        }

        // Union by rank
        if (rank[rootY] > rank[rootX]) {
            root[rootX] = rootY;
            rank[rootY] += rank[rootX];
        } else {
            root[rootY] = rootX;
            rank[rootX] += rank[rootY];
        }

        sets--;  // Decrease number of sets
    }

    // Check if x and y are connected (belong to the same set)
    bool isConnected(int x, int y) {
        return findRootOf(x) == findRootOf(y);
    }
};

